File: public/developer-downloads/examples/java/AureviaProvider.java
// Java 17+ and Jackson (see pom.xml). Backend only; never expose API keys to a browser.
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.net.URI;
import java.net.http.*;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.time.Duration;
import java.time.Instant;
import java.util.*;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

public final class AureviaProvider {
    private static final ObjectMapper JSON = new ObjectMapper();
    private final HttpClient http = HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NEVER)
        .connectTimeout(Duration.ofSeconds(5)).build();
    private final URI base;
    private final String key;

    public AureviaProvider(String baseUrl, String apiKey) {
        base = URI.create(baseUrl.replaceAll("/+$", "") + "/");
        if (!"https".equals(base.getScheme()) || base.getHost() == null || base.getUserInfo() != null ||
            base.getQuery() != null || base.getFragment() != null) throw new IllegalArgumentException("Use the assigned HTTPS base URL.");
        if (apiKey == null || apiKey.isEmpty() || apiKey.contains("\r") || apiKey.contains("\n"))
            throw new IllegalArgumentException("Invalid API key header.");
        key = apiKey;
    }

    public JsonNode request(String method, String path, Object body, String persistedKey) throws Exception {
        if (!path.startsWith("api/provider/v1/") || path.contains("..") || path.contains("\\"))
            throw new IllegalArgumentException("Invalid provider path.");
        var builder = HttpRequest.newBuilder(base.resolve(path)).timeout(Duration.ofSeconds(15))
            .header("X-Provider-API-Key", key).header("Accept", "application/json");
        if (persistedKey != null) builder.header("Idempotency-Key", persistedKey);
        var publisher = HttpRequest.BodyPublishers.noBody();
        if (body != null) {
            builder.header("Content-Type", "application/json");
            publisher = HttpRequest.BodyPublishers.ofString(JSON.writeValueAsString(body), StandardCharsets.UTF_8);
        }
        var response = http.send(builder.method(method, publisher).build(), HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
        if (response.statusCode() < 200 || response.statusCode() >= 300)
            throw new ProviderException(response.statusCode(), response.headers().firstValue("Retry-After").orElse(null));
        return JSON.readTree(response.body()).required("data");
    }

    public JsonNode launch(String authenticatedPlayerId, String returnUrl, String persistedKey, String symbol, String device) throws Exception {
        if (persistedKey == null || persistedKey.isEmpty()) throw new IllegalArgumentException("Persist the launch key before sending; reuse it on retries.");
        if (!Set.of("desktop", "mobile").contains(device)) throw new IllegalArgumentException("Invalid device.");
        return request("POST", "api/provider/v1/launch-url", Map.of("symbol", symbol,
            "external_player_id", authenticatedPlayerId, "currency", "EUR", "lang", "en", "device", device,
            "return_url", returnUrl), persistedKey);
    }

    public static JsonNode verifyCallback(String secret, byte[] rawBody, String timestamp, String signature,
        String requestId, String operation, String operatorId, long now) throws Exception {
        if (secret == null || secret.isEmpty() || operatorId == null || operatorId.isEmpty() ||
            !Set.of("balance", "debit", "credit", "rollback").contains(operation)) throw new IllegalArgumentException("Verifier is not configured.");
        if (timestamp == null || !timestamp.matches("[0-9]{1,12}") ||
            Math.abs(now - Long.parseLong(timestamp)) > 300 || signature == null || !signature.matches("[a-f0-9]{64}"))
            throw new SecurityException("Invalid signature or timestamp.");
        var mac = Mac.getInstance("HmacSHA256");
        mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
        mac.update((timestamp + ".").getBytes(StandardCharsets.US_ASCII));
        if (!MessageDigest.isEqual(mac.doFinal(rawBody), HexFormat.of().parseHex(signature))) throw new SecurityException("Invalid signature.");
        var payload = JSON.readTree(rawBody);
        if (requestId == null || requestId.isEmpty() || !payload.path("request_id").isTextual() ||
            !requestId.equals(payload.path("request_id").asText()) || !operatorId.equals(payload.path("operator_id").asText()) ||
            !operation.equals(payload.path("transaction_type").asText())) throw new SecurityException("Callback identity mismatch.");
        return payload;
    }

    public static final class ProviderException extends Exception {
        public final int status;
        public final String retryAfter;
        ProviderException(int status, String retryAfter) {
            super("Provider HTTP " + status); this.status = status; this.retryAfter = retryAfter;
        }
    }
}
// In your callback controller preserve raw byte[] before parsing, use
// Instant.now().getEpochSecond() for now, then validate the full wallet contract
// and commit ownership checks, ledger mutation and immutable response atomically.

File: public/developer-downloads/examples/AureviaProvider.cs
// .NET 8+. Server-side only. Reuse this client; obtain credentials from a secret store.
using System.Net;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

public sealed class AureviaProvider : IDisposable
{
    private readonly HttpClient http;
    private readonly Uri baseUri;
    private readonly string apiKey;

    public AureviaProvider(string baseUrl, string apiKey)
    {
        baseUri = new Uri(baseUrl.TrimEnd('/') + "/");
        if (baseUri.Scheme != "https" || baseUri.UserInfo != "" || baseUri.Query != "" || baseUri.Fragment != "")
            throw new ArgumentException("Use the assigned HTTPS provider base URL.");
        if (string.IsNullOrEmpty(apiKey) || apiKey.Contains('\r') || apiKey.Contains('\n'))
            throw new ArgumentException("Invalid API key header.");
        this.apiKey = apiKey;
        http = new HttpClient(new HttpClientHandler { AllowAutoRedirect = false, UseCookies = false })
            { Timeout = TimeSpan.FromSeconds(15) };
    }

    public async Task<JsonElement> RequestAsync(HttpMethod method, string path, object? body = null,
        string? idempotencyKey = null, CancellationToken cancellationToken = default)
    {
        if (!path.StartsWith("api/provider/v1/", StringComparison.Ordinal) || path.Contains("..") || path.Contains('\\'))
            throw new ArgumentException("Invalid provider path.");
        using var request = new HttpRequestMessage(method, new Uri(baseUri, path));
        request.Headers.Add("X-Provider-API-Key", apiKey);
        request.Headers.Add("Accept", "application/json");
        if (idempotencyKey != null) request.Headers.Add("Idempotency-Key", idempotencyKey);
        if (body != null) request.Content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");
        using var response = await http.SendAsync(request, cancellationToken);
        if (!response.IsSuccessStatusCode)
            throw new ProviderException(response.StatusCode, response.Headers.RetryAfter?.ToString());
        using var document = JsonDocument.Parse(await response.Content.ReadAsStringAsync(cancellationToken));
        return document.RootElement.GetProperty("data").Clone();
    }

    public Task<JsonElement> LaunchAsync(string authenticatedPlayerId, string returnUrl, string persistedKey,
        string symbol = "wildframes", string device = "desktop", CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrEmpty(persistedKey)) throw new ArgumentException("Persist a key before launching; reuse it with the exact body on retries.");
        if (device != "desktop" && device != "mobile") throw new ArgumentException("Invalid device.");
        return RequestAsync(HttpMethod.Post, "api/provider/v1/launch-url", new {
            symbol, external_player_id = authenticatedPlayerId, currency = "EUR", lang = "en", device, return_url = returnUrl
        }, persistedKey, cancellationToken);
    }

    public static JsonElement VerifyCallback(string secret, byte[] rawBody, string timestamp, string signature,
        string requestId, string operation, string operatorId, long? now = null)
    {
        if (string.IsNullOrEmpty(secret) || string.IsNullOrEmpty(operatorId) ||
            !new[] { "balance", "debit", "credit", "rollback" }.Contains(operation))
            throw new ArgumentException("Wallet verifier is not configured.");
        if (!Regex.IsMatch(timestamp, @"\A[0-9]{1,12}\z") ||
            Math.Abs((now ?? DateTimeOffset.UtcNow.ToUnixTimeSeconds()) - long.Parse(timestamp)) > 300 ||
            !Regex.IsMatch(signature, @"\A[a-f0-9]{64}\z")) throw new CryptographicException("Invalid signature or timestamp.");
        using var hmac = IncrementalHash.CreateHMAC(HashAlgorithmName.SHA256, Encoding.UTF8.GetBytes(secret));
        hmac.AppendData(Encoding.ASCII.GetBytes(timestamp + "."));
        hmac.AppendData(rawBody);
        if (!CryptographicOperations.FixedTimeEquals(hmac.GetHashAndReset(), Convert.FromHexString(signature)))
            throw new CryptographicException("Invalid signature.");
        using var document = JsonDocument.Parse(rawBody);
        var payload = document.RootElement;
        if (string.IsNullOrEmpty(requestId) || payload.GetProperty("request_id").GetString() != requestId ||
            payload.GetProperty("operator_id").GetString() != operatorId || payload.GetProperty("transaction_type").GetString() != operation)
            throw new InvalidOperationException("Callback identity mismatch.");
        return payload.Clone();
    }

    public void Dispose() => http.Dispose();
}

public sealed class ProviderException(HttpStatusCode status, string? retryAfter) : Exception($"Provider HTTP {(int)status}")
{
    public HttpStatusCode Status { get; } = status;
    public string? RetryAfter { get; } = retryAfter;
}

// ASP.NET: read Request.Body as bytes BEFORE JSON deserialization. Pass the four
// x-pp-* header/route values to VerifyCallback. Next validate the entire payload,
// ownership and money fields, and commit wallet + immutable response atomically.
// Signature verification alone is not a wallet implementation.
